co2calculator/calculate.py

```diff
@@ -21,78 +24,149 @@
 detour_df = pd.read_csv(f"{script_path}/../data/detour.csv")
 
 
-def calc_co2_car(
-    distance: Kilometer = None,
-    stops: list = None,
-    passengers: int = None,
-    size: str = None,
-    fuel_type: str = None,
-) -> Tuple[Kilogram, Kilometer]:
-    """
-    Function to compute the emissions of a car trip.
-    :param distance: Distance travelled in km;
-                        alternatively param <stops> can be provided
-    :param stops: List of locations as dictionaries in the form
-                        e.g.,  [{"address": "Im Neuenheimer Feld 348",
-                                "locality": "Heidelberg",
-                                 "country": "Germany"},
-                                 {"country": "Germany",
-                                 "locality": "Berlin",
-                                 "address": "Alexanderplatz 1"}]
-                        can have intermediate stops (> 2 dictionaries within the list)
-                        alternatively param <distance> can be provided
-    :param passengers: Number of passengers in the car (including the person answering the questionnaire),
-                        [1, 2, 3, 4, 5, 6, 7, 8, 9]                             default: 1
-    :param size: size of car
-                        ["small", "medium", "large", "average"]                 default: "average"
-    :param fuel_type: type of fuel the car is using
-                        ["diesel", "gasoline", "cng", "electric", "hybrid", "plug-in_hybrid", "average"]    default: "average"
-    :type distance: float or None
-    :type stops: list[*dict] or None
-    :type passengers: int
-    :type size: str
-    :type fuel_type: str
-    :return: Total emissions of trip in co2 equivalents, total distance of the trip
-    :rtype: tuple[float, float]
+class CalculationError(Exception):
+    """Base exception for all calculation errors"""
+
+
+class InputMissing(CalculationError):
+    """Raised if inputs are not sufficient for calculation"""
+
+
+@enum.unique
+class TransportationMode(str, enum.Enum):
+    """Available transportation modes"""
+
+    CAR = "car"
+    BUS = "bus"
+    TRAIN = "train"
+    PLANE = "plane"
+    FERRY = "ferry"
+
+
+class EmissionRequest(BaseModel):
+    """Generic model for emission calculation request"""
+
+    transportation_mode: TransportationMode
+    distance: Kilometer
+
+
+class Stop(BaseModel):  # Could be renamed to "waypoint"?
+    """Model for a waypoint of a trip"""
+
+    address: str
+    locality: str
+    country: str
```

Comment:
Looks perfectly right.

Few remarks:

1) Regarding you comment if renaming: Models should be as generic as possible if we want to use them across the code base (we put them into co2calculator/models.py then). So, if you want to rename, try to find a name that really suits our problem!
2) If it becomes a more generic model, I could imagine that some of the field are not always applicable. If so, use `typing.Optional` for those. If you want default values, you can add them like so: `address: str = "<your_default_string>"`
3) `Waypoint` should include lat & lng, imho. The model looks more like `Address` to me (which goes into geocoding later). Output of geocoding could be the `Waypoint`!

Keep in mind that my remarks are not related to BaseModel usage. This is more implementation detail and code style. This is where happy discussions start!

Great work!